{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initializing environments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "!pip install ./ONN\n",
    "!ONN init"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Testing installation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2020-09-22 19:13:53.433654: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory\n",
      "2020-09-22 19:13:53.440067: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "usage: ONN [-h] [-i I] [-o O] [-cfg CFG] [-tmp TMP] [-p P] [-otlg OTLG]\n",
      "           [-labels LABELS] [-phylo PHYLO] [-dmax DMAX] [-show] [-from-dir]\n",
      "           [-to-otlg] [-unk] [-db DB] [-in-cm] [-filter-only] [-use-rf] [-C C]\n",
      "           [-split-idx SPLIT_IDX] [-end-idx END_IDX] [-log LOG] [-base BASE]\n",
      "           [-model MODEL] [-ofmt OFMT]\n",
      "           {init,download,map,construct,convert,select,train,transfer,search}\n",
      "\n",
      "The program is designed to help you to transfer Ontology-aware Neural Network model to other source tracking tasks.\n",
      "Feel free to contact us if you have any question.\n",
      "For more information, see Github. Thank you for using Ontology-aware neural network.\n",
      "\n",
      "positional arguments:\n",
      "  {init,download,map,construct,convert,select,train,transfer,search}\n",
      "                        The work mode for ONN program.\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  -i I                  The input file, see input format for each work mode.\n",
      "  -o O                  The output file, see output format for each work mode.\n",
      "  -cfg CFG              The config.ini file.\n",
      "  -tmp TMP              The path to save temperature files.\n",
      "  -p P                  The number of processors to use.\n",
      "  -otlg OTLG            The path to microbiome ontology.\n",
      "  -labels LABELS        The path to npz file (storing labels for the input\n",
      "                        data).\n",
      "  -phylo PHYLO          The phylogeny tree to use, in tsv format.\n",
      "  -dmax DMAX            The max depth of the ontology.\n",
      "\n",
      "construct:\n",
      "  Constructing ontology using microbiome structure \".txt\" file.\n",
      "  Input: microbiome structure \".txt\" file. Output: Constructed microbiome ontology.\n",
      "\n",
      "  -show                 Printing the ontology to stdout.\n",
      "\n",
      "map:\n",
      "  `-from-dir`: Getting mapper file from directory.\n",
      "  Input: The directory to generate mapper file, Output: mapper file.\n",
      "  `-to-otlg`: Mapping source environments to microbiome ontology.\n",
      "  Input: The mapper file, Output: The ontologically arranged labels.\n",
      "\n",
      "  -from-dir             Getting mapper file from directory.\n",
      "  -to-otlg              Mapping source environments to microbiome ontology.\n",
      "  -unk                  Whether to include Unknown source when generating\n",
      "                        labels.\n",
      "\n",
      "convert:\n",
      "  Converting input abundance data to countmatrix at Genus level and generating phylogeny using taxonomic entries involved in the data.\n",
      "  Preparing for feature selection\n",
      "  Input: the input data, Output: RRDM at Genus level\n",
      "\n",
      "  -db DB                The NCBI taxonomy database file to use, in sqlite\n",
      "                        format.\n",
      "  -in-cm                Whether to use the countmatrix as the input format.\n",
      "\n",
      "select:\n",
      "  Selecting features above the threshold. Variance and importance are calculated using Pandas and RandomForestRegressor, respectively.\n",
      "  Input: countmatrix generated by `ONN convert`, Output: selected features and phylogeny (tmp).\n",
      "\n",
      "  -filter-only          Filter features using a selected phylogeny.\n",
      "  -use-rf               Whether to use the randomForest when performing\n",
      "                        selection.\n",
      "  -C C                  The coefficient C in `Threshold = C * mean(stat)`.\n",
      "\n",
      "train:\n",
      "  Training ONN model, the microbiome ontology and properly labeled data must be provided.\n",
      "  Input: samples, in pandas h5 format, output: ONN model\n",
      "\n",
      "  -split-idx SPLIT_IDX  The index to split training and validation samples.\n",
      "  -end-idx END_IDX      The index to split validation and testing samples.\n",
      "  -log LOG              The path to store training history of ONN model.\n",
      "\n",
      "transfer:\n",
      "  Transferring ONN model to fit in a new ontology, The microbiome ontology and properly labeled data must be provided.\n",
      "\n",
      "  -base BASE            The path to base feature extractor model.\n",
      "\n",
      "search:\n",
      "  Searching for source environments of your microbial samples using ONN model.\n",
      "\n",
      "  -model MODEL          The path to ONN model to search against.\n",
      "  -ofmt OFMT            The output format.\n"
     ]
    }
   ],
   "source": [
    "!ONN -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading & preprocessing the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!ONN construct -i tmp/microbiomes.txt -o tmp/ontology.pkl\n",
    "!ONN map -from-dir -i ../data/ -o tmp/mapper.csv\n",
    "!ONN map -to-otlg -otlg tmp/ontology.pkl -i tmp/mapper.csv -o tmp/out/labels.h5 -unk\n",
    "!ONN convert -i tmp/sample-list.txt -conf tmp/conf -o tmp/out/countmatrix_genus.h5 -db ~/.etetoolkit/taxa.sqlite\n",
    "!ONN select -i tmp/conf/phylogeny_by_transformer.csv -cm tmp/out/countmatrix_genus.h5 -o tmp/conf/phylogeny_top1000.csv \\\n",
    "    -top 1000 -labels tmp/out/labels.h5 -dmax 6"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Model training & cross validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!ONN train -i data/matrix-genus-for-combined-C1e-5.h5 -label data/labels-for-combined.h5 \\\n",
    "    -otlg config/ontology-for-combined.pkl -log logs/training-history-for-combined.csv -dmax 6 \\\n",
    "    -o models/model-for-combied -cfg ../../config/config.ini \\\n",
    "    -phylo tmp/phylogeny_selected_using_varianceThreshold_C1e-05_pog.csv -end-idx -1 -split-idx -1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Auto-sorting of Mixed samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!ONN search ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## biome source correction for samples "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
